platform/linux-generic/odp_ml.c

```diff
@@ -0,0 +1,2673 @@
+		return ODP_ML_DATA_TYPE_UINT8;
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT8:
+		return ODP_ML_DATA_TYPE_INT8;
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT16:
+		return ODP_ML_DATA_TYPE_UINT16;
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT16:
+		return ODP_ML_DATA_TYPE_INT16;
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT32:
+		return ODP_ML_DATA_TYPE_INT32;
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT32:
+		return ODP_ML_DATA_TYPE_UINT32;
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64:
+		return ODP_ML_DATA_TYPE_INT64;
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT64:
+		return ODP_ML_DATA_TYPE_UINT64;
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16:
+		return ODP_ML_DATA_TYPE_FP16;
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_BFLOAT16:
+		return ODP_ML_DATA_TYPE_BFP16;
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_DOUBLE:
+		return ODP_ML_DATA_TYPE_FP64;
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_BOOL:
+		/* Fall through */
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_COMPLEX64:
+		/* Fall through */
+	case ONNX_TENSOR_ELEMENT_DATA_TYPE_COMPLEX128:
+		/* Fall through */
+	default:
+		_ODP_ERR("onnx_dtype %d not supported by odp_ml\n", onnx_dtype);
+		return ODP_ML_DATA_TYPE_NONE;
+	}
+}
+
+/* Get the size of given odp_ml_data_type_t in bytes */
+static uint32_t size_of_odp_ml_data_type(odp_ml_data_type_t data_type)
+{
+	switch (data_type) {
+	case ODP_ML_DATA_TYPE_NONE:
+		return 0;
+
+	case ODP_ML_DATA_TYPE_INT8:
+		/* Fall through */
+	case ODP_ML_DATA_TYPE_UINT8:
+		return 1;
+
+	case ODP_ML_DATA_TYPE_INT16:
+		/* Fall through */
+	case ODP_ML_DATA_TYPE_UINT16:
+		/* Fall through */
+	case ODP_ML_DATA_TYPE_FP16:
+		/* Fall through */
+	case ODP_ML_DATA_TYPE_BFP16:
+		return 2;
+
+	case ODP_ML_DATA_TYPE_INT24:
+		/* Fall through */
+	case ODP_ML_DATA_TYPE_UINT24:
+		return 3;
+
+	case ODP_ML_DATA_TYPE_INT32:
+		/* Fall through */
+	case ODP_ML_DATA_TYPE_UINT32:
+		/* Fall through */
+	case ODP_ML_DATA_TYPE_FP32:
+		return 4;
+
+	case ODP_ML_DATA_TYPE_INT64:
+		/* Fall through */
+	case ODP_ML_DATA_TYPE_UINT64:
+		/* Fall through */
+	case ODP_ML_DATA_TYPE_FP64:
+		return 8;
+
+	default:
+		return 0;
+	}
+}
+
+static int get_shape(int64_t dims[], odp_ml_shape_info_t *shape)
+{
+	odp_bool_t first_dyn = false;
+	uint32_t dyn_cnt = 0;
+
+	for (uint32_t i = 0; i < shape->num_dim; i++) {
+		if (dims[i] == 0) {
+			_ODP_ERR("Dimension value: %" PRId64 " must be at least 1\n", dims[i]);
+			return -1;
+		} else if (dims[i] == -1) { /* Symbolic dimension */
+			dyn_cnt++;
+			if (i == 0)
+				first_dyn = true;
+
+			shape->dim[i] = ODP_ML_DIM_DYNAMIC;
+			shape->dim_min[i] = 0; /*unknown*/
+			shape->dim_max[i] = 0; /*unknown*/
+		} else if (dims[i] > 0 && dims[i] < UINT32_MAX) {
+			shape->dim[i] = dims[i];
+			shape->dim_min[i] = dims[i];
+			shape->dim_max[i] = dims[i];
+		} else {
+			_ODP_ERR("Dimension value: %" PRId64 " invalid\n", dims[i]);
+			return -1;
+		}
+	}
+
+	if (dyn_cnt == 0) {
+		shape->type = ODP_ML_SHAPE_STATIC;
+	} else if (dyn_cnt == 1 && first_dyn) {
+		shape->type = ODP_ML_SHAPE_BATCH;
```

Comment:
Does ORT require that batch size is always the first dimension? In general, it could be also some other dimension.

Reply:
ORT doesn't require that. first_dyn may have been related to the multiple dynamic dimensions case, but we don't support that now. I don't see any need for first_dyn. Removed.